### PR TITLE
[mielecloud] Add channels energy and water consumption

### LINKED ISSUE #14414: [mielecloud] Add support for reading water and power consumption
The Miele 3rd Party API provides consumption data for water and power and it is desirable to have these as channels as well.

### PR DESCRIPTION
**Pull request by BjoernLange**
Resolves #14414.

### CODE REVIEW

File: bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/thing/dishwasherDevice.xml [hunk 1]
```diff
@@ -34,6 +34,8 @@
 			<channel id="error_state" typeId="error_state"/>
 			<channel id="info_state" typeId="info_state"/>
 			<channel id="door_state" typeId="door_state"/>
+			<channel id="water_consumption_current" typeId="water_consumption_current"/>
```
Comment (J-N-K) [inline bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/thing/dishwasherDevice.xml:37]: Please make sure to add proper update instructions to allow updating managed things. See https://github.com/openhab/openhab-core/pull/3330

File: bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/i18n/mielecloud_de.properties [hunk 2]
```diff
@@ -238,6 +238,12 @@ channel-type.mielecloud.door_state.description=Gibt an, ob die Tür des Geräts
 channel-type.mielecloud.door_alarm.label=Türalarm
 channel-type.mielecloud.door_alarm.description=Gibt an, ob der Türalarm des Geräts aktiv ist.
 
+channel-type.mielecloud.water_consumption_current.label=Aktueller Wasserverbrauch
+channel-type.mielecloud.water_consumption_current.description=Durch das derzeit laufende Programm verbrauchte Wasser seit Programmstart.
+
+channel-type.mielecloud.energy_consumption_current.label=Aktueller Stromverbrauch
+channel-type.mielecloud.energy_consumption_current.description=Durch das derzeit laufende Programm verbrauchter Strom seit Programmstart.
+
```
Comment (jlaur) [inline bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/i18n/mielecloud_de.properties:246]: Translations have to be provided in Crowdin, see https://www.openhab.org/docs/developer/utils/i18n.html#managing-translations.

**Comment by jlaur** [comment]
@BjoernLange - thanks for the PR. I plan to test it within the next few days. 🙂

File: bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/thing/channelTypes.xml [hunk 3]
```diff
@@ -438,6 +438,22 @@
 		<state readOnly="true"/>
 	</channel-type>
 
+	<channel-type id="water_consumption_current">
+		<item-type>Number:Water</item-type>
```
Comment (jlaur) [inline bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/thing/channelTypes.xml:442]:
```suggestion
		<item-type>Number:Volume</item-type>
```

File: bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/thing/channelTypes.xml [hunk 4]
```diff
@@ -438,6 +438,22 @@
...
+		<label>@text/channel-type.mielecloud.water_consumption_current.label</label>
+		<description>@text/channel-type.mielecloud.water_consumption_current.description</description>
+		<category>Number:Water</category>
```
Comment (jlaur) [inline bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/thing/channelTypes.xml:445]:
```suggestion
		<category>Water</category>
```

or even:

```suggestion
		<category>Water</category>
		<tags>
			<tag>Measurement</tag>
			<tag>Water</tag>
		</tags>
```

File: bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/thing/channelTypes.xml [hunk 5]
```diff
@@ -438,6 +438,22 @@
...
+		<state pattern="%.1f %unit%" readOnly="true"/>
+	</channel-type>
+
+	<channel-type id="energy_consumption_current">
+		<item-type>Number:Energy</item-type>
+		<label>@text/channel-type.mielecloud.energy_consumption_current.label</label>
+		<description>@text/channel-type.mielecloud.energy_consumption_current.description</description>
+		<category>Number:Energy</category>
```
Comment (jlaur) [inline bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/thing/channelTypes.xml:453]:
```suggestion
		<category>Energy</category>
```

or even:
```suggestion
		<category>Energy</category>
		<tags>
			<tag>Measurement</tag>
			<tag>Energy</tag>
		</tags>
```

**Review by jlaur — CHANGES_REQUESTED** [review]
Initial feedback from first phase of testing (linking items to the new channels).

File: bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/i18n/mielecloud_de.properties (hunk 2 above)
Comment (BjoernLange) [reply bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/i18n/mielecloud_de.properties:246]: Will do so, but what does this mean for the translation files? Should I leave it as is now, remove the file or remove the new entries?

File: bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/thing/dishwasherDevice.xml (hunk 1 above)
Comment (BjoernLange) [reply bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/thing/dishwasherDevice.xml:37]: Is there any developer documentation on how to do this apart from the examples from the PR?

File: bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/i18n/mielecloud_de.properties (hunk 2 above)
Comment (jlaur) [reply bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/i18n/mielecloud_de.properties:246]: Only remove the new entries, that will avoid collisions with Crowdin being upstream. The new entries can then be added in Crowdin when the PR is merged, and the proof-reading process will be utilized there.

File: bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/thing/dishwasherDevice.xml (hunk 1 above)
Comment (mhilbush) [reply bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/thing/dishwasherDevice.xml:37]: Not yet. But you also could have a look at the PR I just submitted for some channels I added to the sleepiq binding.

https://github.com/openhab/openhab-addons/pull/14460

Comment (BjoernLange) [reply bundles/org.openhab.binding.mielecloud/src/main/resources/OH-INF/thing/dishwasherDevice.xml:37]: Is it required to add the `thingTypeVersion` property for each thing type with update instructions?

```
		<properties>
			<property name="thingTypeVersion">1</property>
		</properties>
```

Does this serve the purpose of preventing updates on newly created things?

**Comment by jlaur** [comment]
@BjoernLange - in case you haven't noticed, the integration tests are failing.

**Review by J-N-K — COMMENTED** [review]
LGTM regarding the update instructions.

**Comment by BjoernLange** [comment]
I suspect the integration tests are failing because the things that now have update instructions are missing the `thingTypeVersion` property. Is there a way to instantiate things as they are defined in the XML thing type description from OH-INF during test setup? Otherwise I will just attach the property for the three affected handler types by hand. Maybe you can give me a hint @J-N-K?

**Comment by BjoernLange** [comment]
I fixed the integration tests, would you mind having another look?

**Review by jlaur — CHANGES_REQUESTED** [review]
Besides that SAT finding, LGTM.

I have tested on my productive system by backporting to 3.4.x. Unfortunately I haven't been able to make it work - both new channels stay **UNDEF** for my washing machine. Nothing of interest is logged, at least I see no payloads or other indicators with **trace** level. Have you been able to test successfully yourself, or do you have confirmation from other users? My washing machine is connected through XGW 3000, and I can see the consumption data in the Miele@home app.

**Comment by jlaur** [comment]
@BjoernLange - okay, I added some logging of the payload:

```json
	"xxxxxxxxxxxx": {
		"ident": {
			"type": {
				"key_localized": "Device type",
				"value_raw": 1,
				"value_localized": "Washing machine"
			},
			"deviceName": "Vaskemaskine",
			"protocolVersion": 2,
			"deviceIdentLabel": {
				"fabNumber": "xxxxxxxxxxxx",
				"fabIndex": "10",
				"techType": "WMV960WPS",
				"matNumber": "xxxxxxxx",
				"swids": [
					"3918",
					"25204",
					"25117",
					"20163",
					"3944",
					"20242",
					"25042",
					"25043",
					"25026",
					"2971",
					"20166",
					"2971",
					"20166"
				]
			},
			"xkmIdentLabel": {
				"techType": "XKM3000Z",
				"releaseVersion": "01.16"
			}
		},
		"state": {
			"ProgramID": {
				"value_raw": 1,
				"value_localized": "Cottons",
				"key_localized": "Program name"
			},
			"status": {
				"value_raw": 5,
				"value_localized": "In use",
				"key_localized": "status"
			},
			"programType": {
				"value_raw": 1,
				"value_localized": "Own programme",
				"key_localized": "Program type"
			},
			"programPhase": {
				"value_raw": 261,
				"value_localized": "Rinsing",
				"key_localized": "Program phase"
			},
			"remainingTime": [
				0,
				10
			],
			"startTime": [
				0,
				0
			],
			"targetTemperature": [
				{
					"value_raw": 3000,
					"value_localized": 30.0,
					"unit": "Celsius"
				}
			],
			"coreTargetTemperature": [],
			"temperature": [],
			"coreTemperature": [],
			"signalInfo": false,
			"signalFailure": false,
			"signalDoor": false,
			"remoteEnable": {
				"fullRemoteControl": true,
				"smartGrid": false,
				"mobileStart": false
			},
			"ambientLight": null,
			"light": null,
			"elapsedTime": [
				1,
				26
			],
			"spinningSpeed": {
				"unit": "rpm",
				"value_raw": 1600,
				"value_localized": "1600",
				"key_localized": "Spin speed"
			},
			"dryingStep": {
				"value_raw": null,
				"value_localized": "",
				"key_localized": "Drying level"
			},
			"ventilationStep": {
				"value_raw": null,
				"value_localized": "",
				"key_localized": "Fan level"
			},
			"plateStep": [],
			"ecoFeedback": null,
			"batteryLevel": null
		}
	},
```

so it seems this information is simply not available for me. The Miele@home app must use some other API to get this information.

**Review by jlaur — APPROVED** [review]
LGTM

**Pull request merged**